packages/runtime/container-runtime/src/containerRuntime.ts

```diff
@@ -1679,14 +1699,46 @@ export class ContainerRuntime extends TypedEventEmitter<IContainerRuntimeEvents>
         }
     }
 
-    public async createDataStore(pkg: string | string[]): Promise<IFluidRouter> {
-        return this._createDataStore(pkg, false /* isRoot */);
+    private async createDataStoreCore(
+        pkg: string | string[],
+        isRoot: boolean,
+        id: string,
+        props?: any,
+    ): Promise<IFluidDataStoreChannel> {
+        const fluidDataStore = await this._createDataStore(pkg, isRoot, id, props);
+        if (isRoot) {
+            fluidDataStore.bindToContext();
+        }
+
+        return fluidDataStore;
+    }
+
+    public async createDataStore(pkg: string | string[]): Promise<IDataStore> {
+        const internalId = uuid();
+        return channelToDataStore(
+            await this.createDataStoreCore(pkg, false /* isRoot */, internalId),
+            internalId,
+            this,
+            this.mc.logger);
+    }
+
+    private async createAndAliasDataStore(pkg: string | string[], alias: string, props?: any): Promise<IDataStore> {
+        const internalId = uuid();
+        const dataStore = await this.createDataStoreCore(pkg, false /* isRoot */, internalId, props);
+        dataStore.bindToContext();
+        const aliasedDataStore = channelToDataStore(dataStore, internalId, this, this.mc.logger);
+        const result = await aliasedDataStore.trySetAlias(alias);
+        if (!result) {
+            throw new GenericError("dataStoreAliasConflict");
+        }
+
+        return aliasedDataStore;
     }
 
     public async createRootDataStore(pkg: string | string[], rootDataStoreId: string): Promise<IFluidRouter> {
-        const fluidDataStore = await this._createDataStore(pkg, true /* isRoot */, rootDataStoreId);
-        fluidDataStore.bindToContext();
-        return fluidDataStore;
+        return this._aliasingEnabled === true && this.attachState === AttachState.Attached ?
```

Comment:
Testing for attached is almost always wrong. I believe you want this.attachState !== AttachState.Detached.

Reply:
Same below